rdbms/src/main/java/org/spine3/server/storage/jdbc/table/entity/RecordTable.java

```diff
@@ -0,0 +1,141 @@
+import static org.spine3.server.storage.jdbc.Sql.Type.BOOLEAN;
+import static org.spine3.server.storage.jdbc.Sql.Type.UNKNOWN;
+
+/**
+ * A table for storing the {@link EntityRecord entity records}.
+ *
+ * <p>Used in the {@link org.spine3.server.storage.jdbc.entity.JdbcRecordStorage}.
+ *
+ * @author Dmytro Dashenkov.
+ */
+public class RecordTable<I> extends EntityTable<I, EntityRecord, RecordTable.Column> {
+
+    private final RecordStorageQueryFactory<I> queryFactory;
+
+    public RecordTable(Class<Entity<I, ?>> entityClass,
+                       DataSourceWrapper dataSource) {
+        super(entityClass, dataSource);
+        queryFactory = new RecordStorageQueryFactory<>(dataSource, entityClass, log());
+    }
+
+    @Override
+    public Column getIdColumnDeclaration() {
+        return Column.id;
+    }
+
+    @Override
+    protected Class<Column> getTableColumnType() {
+        return Column.class;
+    }
+
+    @Override
+    protected QueryFactory<I, EntityRecord> getQueryFactory() {
+        return queryFactory;
+    }
+
+    public boolean markDeleted(I id) {
+        return queryFactory.newMarkDeletedQuery(id)
+                           .execute();
+    }
+
+    public boolean markArchived(I id) {
+        return queryFactory.newMarkArchivedQuery(id)
+                           .execute();
+    }
+
+    public Map<?, EntityRecord> read(Iterable<I> ids, FieldMask fieldMask) {
+        try {
+            final Map<?, EntityRecord> recordMap = queryFactory.newSelectBulkQuery(ids, fieldMask)
+                                                               .execute();
+            return recordMap;
+        } catch (SQLException e) {
+            throw new DatabaseException(e);
+        }
+    }
+
+
+    public void write(Map<I, EntityRecord> records) {
+        // Map's initial capacity is maximum, meaning no records exist in the storage yet
+        final Map<I, EntityRecord> newRecords = new HashMap<>(records.size());
+
+        for (Map.Entry<I, EntityRecord> unclassifiedRecord : records.entrySet()) {
+            final I id = unclassifiedRecord.getKey();
+            final EntityRecord record = unclassifiedRecord.getValue();
+            if (containsRecord(id)) {
+                queryFactory.newUpdateQuery(id, record)
+                            .execute();
+            } else {
+                newRecords.put(id, record);
+            }
+        }
+        queryFactory.newInsertEntityRecordsBulkQuery(newRecords)
+                    .execute();
+    }
+
+    public Map<I, EntityRecord> readAll(FieldMask fieldMask) {
+        final SelectBulkQuery<I> query = queryFactory.newSelectAllQuery(fieldMask);
+        try {
+            final Map<I, EntityRecord> result = query.execute();
+            return result;
+        } catch (SQLException e) {
+            throw new DatabaseException(e);
+        }
+    }
+
+    public enum Column implements TableColumn, Cloneable {
+
+        id(UNKNOWN),
```

Comment:
Can we have a type name, which says that the type would be determined on run-time. The thing is that names like `UNKNOWN` are reserved for errors in Protobuf enums. This does look like an error.